src/stream/src/executor/backfill/arrangement_backfill.rs

```diff
@@ -159,13 +175,19 @@ where
                 // It is finished, so just assign a value to avoid accessing storage table again.
                 false
             } else {
-                let snapshot = Self::snapshot_read_per_vnode(
-                    &upstream_table,
-                    backfill_state.clone(), // FIXME: temporary workaround... How to avoid it?
-                    &mut builders,
-                );
-                pin_mut!(snapshot);
-                snapshot.try_next().await?.unwrap().is_none()
+                let snapshot_is_empty = {
+                    let snapshot = Self::snapshot_read_per_vnode(
+                        &upstream_table,
+                        backfill_state.clone(), // FIXME(kwannoel): use mutable reference instead.
+                        &mut builders,
+                    );
+                    pin_mut!(snapshot);
+                    let next = snapshot.try_next().await?;
+                    next.is_none()
+                };
+                let builder_is_empty = builders.iter().all(|b| b.is_empty());
+                builders.iter_mut().for_each(|b| b.clear());
+                snapshot_is_empty && builder_is_empty
```

Comment:
Maybe we can get rid of this empty snapshot check? I remember we used to remove it from the no shuffle backfill.

Reply:
Fixed